components/script/dom/bluetooth.rs

```diff
@@ -40,12 +56,155 @@ impl Bluetooth {
     }
 }
 
+impl Clone for StringOrUnsignedLong {
+    fn clone(&self) -> StringOrUnsignedLong {
+        match self {
+            &StringOrUnsignedLong::String(ref s) => StringOrUnsignedLong::String(s.clone()),
+            &StringOrUnsignedLong::UnsignedLong(usl) => StringOrUnsignedLong::UnsignedLong(usl),
+        }
+    }
+}
+
+impl Clone for BluetoothScanFilter {
+    fn clone(&self) -> BluetoothScanFilter {
+        match self.services {
+            None => BluetoothScanFilter {
+                name: self.name.clone(),
+                namePrefix: self.namePrefix.clone(),
+                services: None,
+            },
+            Some(ref vec) => BluetoothScanFilter {
+                name: self.name.clone(),
+                namePrefix: self.namePrefix.clone(),
+                services: Some(vec.clone()),
+            },
+        }
+    }
+}
+
+impl Clone for RequestDeviceOptions {
+    fn clone(&self) -> RequestDeviceOptions {
+        match self.optionalServices {
+            None => RequestDeviceOptions {
+                filter: self.filter.clone(),
+                optionalServices: None,
+            },
+            Some(ref vec) => RequestDeviceOptions {
+                filter: self.filter.clone(),
+                optionalServices: Some(vec.clone()),
+            }
+        }
+    }
+}
+
+fn canonicalize_filter(filter: &BluetoothScanFilter, global: GlobalRef)
+    -> Fallible<BluetoothScanfilter> {
+        if !(filter.services.is_some() ||
```

Comment:
this indentation looks beautiful, but I'm not sure if the style allows it. maybe unindent by one, and see if it looks good